rfcs/convergence/simplify-prop-merging.md

```diff
@@ -0,0 +1,326 @@
+# RFC: Simplify prop merging with slots
+
+---
+
+@layershifter @ling1726
+
+## Summary
+
+This RFC proposes a set of changes in three major steps:
+
+1. Fixing the use of slots in `mergeProps` and usage of the `as` prop, so that JSX elements in slots can be configured using shorthand `as` prop.
+2. Stricter typing for shorthand slots that render native HTML tags with the goal of improving type safety internally and for consumers.
+3. A logical conclusion of the two above steps which results completely removing `mergeProps` in favour of object spreading over deep merge.
+
+The later sections will go into the details of what each step will do and the potential benefits.
+
+## Background
+
+`mergeProps` is a layer over quite a complicated deep merging functionality is built to handle object merges with React elements properties, this results in quite complicated typings. Recently efforts have been made to make the usage safer from a typescript point of view thanks to concerned engineers who give up some of their project cycles.
+
+This proposal aims to tackle the following problems more fundamentally in the design of shorthands and `mergeProps` itself:
+
+- `as` prop is broken when passed to JSX elements through shorthand slots
+- No type safety declaring HTML attributes with `as`
+- Unsafe generic typings for shorthand, that require complex helper types above the internal implementation to make it work
+- Deep merges are costly
+- className merges are not done with `useAx`, which handles scoped directional styles and dedupes atomic declarations
+- Functions are not really merged but simply replaced similar to native object spreading
```

Comment:
````suggestion
- Functions are not really merged but simply replaced similar to native object spreading

```tsx
const props = { button: { customHandleClick /* a custom handler */} };
const state = mergeProps(
  { button: { handleClick /* a require handler by our component */ } },
   resolveShorthandProps(props),
);
// Results in => { button: { customHandleClick /* 💣 handlers are not merged */ } }
```

````

Reply:
It might be unclear what it is covered there, as we discussed this case in offline I think that it will be great to add a snippet that explains the problem. WDYT?